### PR TITLE
fixes #11506 - import rhsm facts to hosts

### PR DESCRIPTION
**Pull request by thomasmckay**

### CODE REVIEW

**Comment by stbenjam** [comment]
I know it's a WIP but it might be worth considering writing and registering a FactImporter instead:

  https://github.com/theforeman/foreman_salt/blob/master/app/services/foreman_salt/fact_importer.rb
  https://github.com/theforeman/foreman_salt/blob/master/lib/foreman_salt/extensions.rb#L2

 
  https://github.com/theforeman/foreman_chef/blob/master/app/models/foreman_chef/fact_importer.rb

**Comment by stbenjam** [comment]
And since subman provides some structured/nested facts, to use the limited support already available for it.

File: app/controllers/katello/api/v2/systems_controller.rb [hunk 1]
```diff
@@ -139,6 +139,15 @@ def create
...
       system_params = system_params(params)
+
+      if params[:facts] && params[:facts]['network.hostname']
+        host = Host.where(:name => params[:facts]['network.hostname'], :organization_id => @system.organization.id).first
+      elsif @system.facts['network.hostname']
+        host = Host.where(:name => @system.facts['network.hostname'], :organization_id => @system.organization.id).first
+      else
+        host = Host.where(:name => @system.name, :organization_id => @system.organization.id).first
+      end
+      system_params.merge!(:host_id => host.try(:id))
```
Comment (jlsherrill) [inline app/controllers/katello/api/v2/systems_controller.rb]: You might just use @system.foreman_host

File: app/lib/actions/katello/system/facts.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+          fact_value = input[:host]
+          fact = ::Katello::FactName.where(:name => fact_name).first
+          if fact.nil?
+            fact = ::Katello::FactName.new(:name => fact_name)
+            fact.save
+          end
+          host.fact_values.create!(:fact_name_id => fact.id, :value => fact_value)
+
+        rescue Exception => e
+          x = e
```
Comment (jlsherrill) [inline app/lib/actions/katello/system/facts.rb]: why the x = e ?  why not just let the exception bubble up?

Comment (thomasmckay) [inline app/lib/actions/katello/system/facts.rb]: That was debug code, thanks.

**Comment by thomasmckay** [comment]
@stbenjam I did look at the salt fact importer before I started but it seemed overly complex for the basics of sub-mgr facts. Yes, there is grouping represented in the keys (period separated) but storing it other than strings seems unnecessary.

Are there use cases with puppet or salt facts where the saving of structured data is critical to functionality of use cases?

File: app/controllers/katello/api/v2/systems_controller.rb (hunk 1 above)
Comment (thomasmckay) [inline app/controllers/katello/api/v2/systems_controller.rb]: Removed this as it was outside the scope of the PR (but useful for testing). Functionality won't be needed after host unification anyway.

**Comment by thomasmckay** [comment]
just :ship: it!

**Comment by stbenjam** [comment]
> @stbenjam I did look at the salt fact importer before I started but it seemed overly complex for the basics of sub-mgr facts. Yes, there is grouping represented in the keys (period separated) but storing it other than strings seems unnecessary.

> Are there use cases with puppet or salt facts where the saving of structured data is critical to functionality of use cases?

IMHO Katello should do things the way other plugins do, even if its too complicated.  You don't have to use the nested facts, but I think you should.

`net.interface.docker0.ipv4_address` should represent a nested fact though in Foreman, and displayed as a hierarchy: net::interface::docker0::ipv4_address, which would let you query all the facts at the net::interface::docker0 level, for example.

Even if you don't want to do that, the implementation code should go into a FactImporter class so at least that's the same.

**Comment by stbenjam** [comment]
Also, subman already provides the facts in that format, I don't think it'd be difficult to use Foreman's nested facts format.  Most of the code in the Salt and Chef importers involves traversing a hash and creating the appropriate structure.

**Comment by thomasmckay** [comment]
I'll open a new PR with fancier facts.

**Pull request closed without merging**